### ISSUE TITLE
Types not recognized in records

### ISSUE THREAD

**Issue by 6d77** [souffle-lang/souffle#810]
The following code does not compile in Souffle:

```prolog
.comp Component {
    .number_type N1
    .type T1
    
    .type Record = [
        n1: N1,
        t1: T1
    ]
    
    .decl record(record: Record)
    
    record([1, "1"]).
}

.init comp = Component
```

with the error messages

```
Error: Undefined type N1 in definition of field n1 in unknown source location.
Error: Undefined type T1 in definition of field t1 in unknown source location.
Error: Number constant (type mismatch) in file RecordInComp.dl at line 12
    record([1, "1"]).
------------^----------
Error: Symbol constant (type mismatch) in file RecordInComp.dl at line 12
    record([1, "1"]).
---------------^-------
4 errors generated, evaluation aborted
```

Substituting `N1` by `number` and `T1` by `symbol` in the record definition works well.

**Comment by b-scholz** [comment]
Scoping is not working well with the type expansion. Indeed this is a bug. Here is a simple workaround
```
.number_type N1
.symbol_type T1
.comp Component {

    .type Record = [
        n1: N1,
        t1: T1
    ]

    .decl record(record: Record)

    record([1, "1"]).
}

.init comp = Component
```

**Comment by olligobber** [comment]
This is an issue with unions as well as records, as the following does not compile:
```
.comp Component {
    .type T1
    .type Un = T1 | number
    .decl test(x : Un)
    .decl symbol(x : T1)
    symbol("A").
    test(x) :- symbol(x).
}
.init comp = Component
```
but a similar work-around as in records works:
```
.type T1
.comp Component {
    .type Un = T1 | number
    .decl test(x : Un)
    .decl symbol(x : T1)
    symbol("A").
    test(x) :- symbol(x).
}
.init comp = Component
```

**Comment by azreika** [comment]
After a lot of chasing, the problem seems to be that the typenames within record types aren't updated when a component is instantiated.

Example:
```
.comp Component {
    .type N1
    .type Record = [n1: N1]
    .decl record(record: Record)
    record(["z"]).
}
.init comp = Component
```

During the type analysis, `.type N1` is transformed into `.type comp.N1`, but `.type Record = [n1: N1]` is /not/ changed to `.type Record = [n1: comp.N1]`, as it should be.

I'll pull up a fix soon.

**Comment by mmcgr** [comment]
Fixed by #835.
